enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,310 @@
+   ...
+</label>
+```
+
+Following is a diagram which identifies the workflow of a message through fluentd as captured by the above configuration.
+
+![https://docs.google.com/drawings/d/18FD4c1pBz4e8eJIH37yUwM40ew69g8h1cqPEfOLsuyY/edit?usp=sharing](images/log-forwarding-message-flow.png)
+
+### Cluster Logging Operator
+* The operator generates output configuration for the collector
+* The operator specifies in the collector daemonset the secrets needed to communicate to each endpoint:
+
+   `/var/run/ocp-collector/secrets/<secretName>/{cert,key,cacert}`
+
+    where:
+
+    * `secretName` is the name of a secret in the `openshift-logging` namespace.
+    * `cert`, `key`, `cacert` are files in the secret
+#### Collector
+* Collectors will be modified to be remove endpoint config specific logic from the start script ; configuration is assumed to be correct and used as provided by the cluster-logging-operator
+* Extract all configuration from after the source config point to final destination into the collector configuration.
+
+### Test Plan
+#### Unit testing
+* Log forwarding will add unit tests to provide adequate coverage for all changes
+* BDD unit testing will be added to unit testing to make tests goals more explicit,  readable, and obvious
+#### Integration and E2E tests
+* Test no regressions after upgrade when log forwarding exists
+* Tests to verify there are no regressions in functionality between pre and post logforwarding
+* Tests to verify log forwarding is writing logs to an Elasticsearch instance not managed by cluster logging
+* Tests to verify log forwarding is writing logs to an Elasticsearch instance managed by cluster logging
+* Tests to verify log forwarding is writing logs to a Fluentd instance via secure forward that is not managed by cluster logging
+
+### Graduation Criteria
+
+##### Tech Preview
+
+* Ability to store logs to the cluster managed log aggregator without defining log forwarding (e.g. status quo)
+* Explicitly define and document log sources and their meanings (e.g. `logs.apps`, `logs.infa`)
+* Explicitly define supported targets (e.g. `elasticsearch`, `fluentd`)
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA 
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Concrete API
+- Additional supported targets (e.g. `Splunk`, `Graylog`, `kafka`)
+
+### Upgrade / Downgrade Strategy
+
+* Upgrades should be seemless to users of cluster logging deployments managed by the cluster-logging-operator.  The current implementation is a static file mounted into the collector daemonset.  Currently user's must explicitly identify a deployment as 'Unmanaged' in order to make changes to the collector's configuration
+* Downgrades should be seemless for the same reasons identified for upgrades.
```

Comment:
even if they are using the pipelines feature at the time they downgrade?

Reply:
How exactly should we express this.  The default is to always ship logs to the managed ES. Admins who do nothing should see no difference even though this feature would still be in use after upgrade.  User's who use the feature and then downgrade would obviously no longer see logs arriving at their specified destination.  They would have to go and hack fluent's config manually to make logs flow again.

Reply:
i think i'd just call it out exactly like that. 

what would it mean for logs getting to the managed ES after downgrade?  would keep working?  What if they had disabled forwarding to managed ES and then they downgrade?  will it start forwarding since the older logging stack won't understand the config that disabled it?

not saying we have to solve all these scenarios, but we need to at least outline expectations.

Reply:
I am making an update to discourage downgrade as this is generally a scenario we need to understand more broadly.  Elasticsearch could run into trouble if we try to downgrade a version because of the data migration it takes during upgrade.